pkg/apiserver/sync/crd.go

```diff
@@ -0,0 +1,547 @@
+
+	// delete the components that not belongs to the new app
+	for _, entity := range originPolicies {
+		plc := entity.(*model.ApplicationPolicy)
+		// we only compare for policies that automatically generated by sync process, and the policy should not be ref ones.
+		if plc.Creator != model.AutoGenPolicy {
+			continue
+		}
+		if !utils.StringsContain(readyToDelete, plc.Name) {
+			continue
+		}
+		if err := ds.Delete(ctx, plc); err != nil {
+			if errors.Is(err, datastore.ErrRecordNotExist) {
+				continue
+			}
+			log.Logger.Warnf("delete policy %s for app %s failure %s", plc.Name, appPrimaryKey, err.Error())
+		}
+	}
+
+	// add or update new app's policies for datastore
+	for _, plc := range expPolicies {
+		if utils.StringsContain(readyToAdd, plc.Name) {
+			err = ds.Add(ctx, plc)
+		} else {
+			err = ds.Put(ctx, plc)
+		}
+		if err != nil {
+			log.Logger.Warnf("convert policy %s for app %s into datastore failure %s", plc.Name, utils2.Sanitize(appPrimaryKey), err.Error())
+			return err
+		}
+	}
+	return nil
+}
+
+// StoreWorkflow will sync workflow application CR to datastore, it updates the only one workflow from the application with specified name
+func StoreWorkflow(ctx context.Context, targetApp *model.DataStoreApp, ds datastore.DataStore) error {
+	err := ds.Get(ctx, &model.Workflow{AppPrimaryKey: targetApp.Name, Name: targetApp.Workflow.Name})
+	if err == nil {
+		// it means the record already exists, update it
+		return ds.Put(ctx, targetApp.Workflow)
+	}
+	if !errors.Is(err, datastore.ErrRecordNotExist) {
+		// other database error, return it
+		return err
+	}
+	return ds.Add(ctx, targetApp.Workflow)
+}
+
+// StoreTargets will sync targets from application CR to datastore
+func StoreTargets(ctx context.Context, targetApp *model.DataStoreApp, ds datastore.DataStore) error {
+	for _, t := range targetApp.Targets {
+		err := ds.Get(ctx, t)
+		if err == nil {
+			continue
+		}
+		if !errors.Is(err, datastore.ErrRecordNotExist) {
+			// other database error, return it
+			return err
+		}
+		if err = ds.Add(ctx, t); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Store2UXAuto will sync application CR to storage of VelaUX automatically
+func Store2UXAuto(ctx context.Context, cli client.Client, targetApp *v1beta1.Application, ds datastore.DataStore) error {
+	sot := CheckSoTFromCR(targetApp)
+
+	// This is a double check to make sure the app not be converted and un-deployed
+	sot = CheckSoTFromAppMeta(ctx, ds, targetApp.Name, sot)
+
+	switch sot {
+	case FromUX, FromInner:
+		// we don't sync if the application is not created from CR
+		return nil
+	default:
+	}
+
+	dsApp, err := ConvertApp2DatastoreApp(ctx, cli, targetApp)
+	if err != nil {
+		log.Logger.Errorf("Convert App to data store err %v", err)
+		return err
+	}
+
+	if err = StoreAppMeta(ctx, dsApp, ds); err != nil {
+		log.Logger.Errorf("Store App Metadata to data store err %v", err)
```

Comment:
What if any one of the StoreXXX fails? Partial information is stored but others are not. Should there be rollback? Or is it idempotent?